File: Algomonster/1-Foundation/9-Sorting/count_swaps_bubble.py
def count_swaps_bubble_sort(arr: list[int]) -> int:
    swaps_count = 0

    for j in range(0, -len(arr), -1):
        old_swaps_count = swaps_count

        for i in range(len(arr) - 1):
            if arr[i+1] < arr[i]:
                arr[i], arr[i+1] = arr[i+1], arr[i]
                swaps_count += 1
        
        if swaps_count == old_swaps_count:
            break

    print(arr)
    return swaps_count

print(count_swaps_bubble_sort([1, 2, 3, 4, 5]))     # 0
print(count_swaps_bubble_sort([5, 4, 3, 2, 1]))     # 10
print(count_swaps_bubble_sort([3, 1, 4, 2]))        # 3

File: Algomonster/1-Foundation/9-Sorting/count_swaps_bubble.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int count_swaps_bubble_sort(std::vector<int> &arr) {
    int n = arr.size();
    int swap_count = 0;

    // sorted = first sorted ind
    for (int sorted = n; sorted > 0; sorted--) {
        bool swapped = false;
        for (int index = 0; index < n - 1; index++) {
            if (arr[index] > arr[index + 1]) {
                std::swap(arr[index], arr[index + 1]);
                swap_count++;
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }
    return swap_count;
}

int main() {
    std::vector<int> numbers = {64, 34, 25, 12, 22, 11, 90};
    int count = count_swaps_bubble_sort(numbers);

    for (int num : numbers) {
        std::cout << num << " ";
    }
    // Output: 11 12 22 25 34 64 90

    std::cout << "\nFinal Count: " << count << std::endl;
    return 0;
}